src/main/java/gregtech/api/util/virtualregistry/entries/VirtualTank.java

```diff
@@ -0,0 +1,186 @@
+
+    protected static final String CAPACITY_KEY = "capacity";
+    protected static final String FLUID_KEY = "fluid";
+    private static final int DEFAULT_CAPACITY = 64000; // 64B
+
+    @Nullable
+    private FluidStack fluidStack = null;
+    private int capacity;
+    private final IFluidTankProperties[] props = new IFluidTankProperties[] {
+            createProperty(this)
+    };
+
+    public VirtualTank(int capacity) {
+        this.capacity = capacity;
+    }
+
+    public VirtualTank() {
+        this(DEFAULT_CAPACITY);
+    }
+
+    @Override
+    public EntryTypes<VirtualTank> getType() {
+        return EntryTypes.ENDER_FLUID;
+    }
+
+    @Override
+    public FluidStack getFluid() {
+        return this.fluidStack;
+    }
+
+    public void setFluid(FluidStack fluid) {
+        this.fluidStack = fluid;
+    }
+
+    @Override
+    public int getFluidAmount() {
+        return fluidStack == null ? 0 : fluidStack.amount;
+    }
+
+    @Override
+    public int getCapacity() {
+        return this.capacity;
+    }
+
+    @Override
+    public FluidTankInfo getInfo() {
+        return new FluidTankInfo(this);
+    }
+
+    @Override
+    public IFluidTankProperties[] getTankProperties() {
+        return this.props;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof VirtualTank other)) return false;
+        if (this.fluidStack == null && other.fluidStack == null)
+            return super.equals(o);
+        if (this.fluidStack == null || other.fluidStack == null)
+            return false;
+        if (this.fluidStack.isFluidStackIdentical(other.fluidStack))
+            return super.equals(o);
+
+        return false;
+    }
+
+    @Override
+    public NBTTagCompound serializeNBT() {
+        var tag = super.serializeNBT();
+        tag.setInteger(CAPACITY_KEY, this.capacity);
+
+        if (this.fluidStack != null)
+            tag.setTag(FLUID_KEY, this.fluidStack.writeToNBT(new NBTTagCompound()));
+
+        return tag;
+    }
+
+    @Override
+    public void deserializeNBT(NBTTagCompound nbt) {
+        super.deserializeNBT(nbt);
+        this.capacity = nbt.getInteger(CAPACITY_KEY);
+
+        if (nbt.hasKey(FLUID_KEY))
+            setFluid(FluidStack.loadFluidStackFromNBT(nbt.getCompoundTag(FLUID_KEY)));
+    }
+
+    @Override
+    public int fill(FluidStack fluidStack, boolean doFill) {
+        if (fluidStack == null || fluidStack.amount <= 0 ||
+                (this.fluidStack != null && !fluidStack.isFluidEqual(this.fluidStack)))
+            return 0;
+
+        int fillAmt = Math.min(fluidStack.amount, getCapacity() - this.getFluidAmount());
+
+        if (doFill) {
+            var fluid = this.fluidStack == null ? null : this.fluidStack.copy();
+
+            if (fluid == null) {
+                fluid = new FluidStack(fluidStack, fillAmt);
+            } else {
+                fluid.amount += fillAmt;
+            }
```

Comment:
This logic can be cleaned up, particularly the ternary. Also, why do we need to copy this.fluidstack here? We currently change the count and overwrite the old one, instead of directly modifying the current one.

Reply:
i think i had a reason for copying, but not anymore.